README
========================================
 _ _ _      ____ ____  ____             
| (_) |__  / ___/ ___||  _ \  _     _   
| | | '_ \| |   \___ \| |_) || |_ _| |_ 
| | | |_) | |___ ___) |  __/_   _|_   _|
|_|_|_.__/ \____|____/|_|    |_|   |_|  
========================================


libCSP++ is a small library whose purpose is allowing the programmer who has to
manage constraint-related problems (CSP == Constraint Satisfaction Problem), in
artificial intelligence or any other application, to do it easily, just caring
about how to express the constraints of his own problem, not how to write an
algorithm that verifies those constraints. The programmer only has to focus
three parameters: the variables of the problem (e.g., in a sudoku resolution 
algorithm, each cell of the game may represent a variable), the domains of these
variables (e.g., in a sudoku, each cell may assume values from 1 to 9) and the
constraints of the problem, simply provided like one or more boolean functions
(always in a sudoku, the constraints are represented by a boolean function
returning true if all the variables in the same row, column and cell have
different values, false otherwise). Once focused these features, the programmer
just has to provide them to build a CSP object, without caring of the algorithm
to check the satisfiability of his problem, and so pay his attention to *HIS*
own algorithm.


INSTALLATION:

Just type make install. The files will be copied, by default, to /usr/local.
Change the Makefile if you want them in a different location. No additional
dependency is required. Only a C++ compiler.


USAGE:

Just include <csp++/csp++.h> in your source files using libCSP++, possibly
adding the directory where csp++ directory is located to your include list (for
example, using g++ and supposing you installed csp++ to /usr/local, you would
append -I/usr/local/include to your compiler command line).


DOCUMENTATION:

You can find doxygen-generated documentation, both in HTML and LaTeX format, in
the doc/ directory of the project. If you modify some of the sources files of
the library and want to change the documentation, just move to csp++ directory
and launch `doxygen' command again.


EXAMPLES:

Two examples of usage of the library are provided.

- fourcolours.cpp is a simple application whose purpose is to interactively ask the
user to enter the colours for some European countries. The constraint is that two
adjacent countries on the map cannot have the same colour, and the domain of
each variable (country) is the set of colours it can assume (a set of 4 colours
is used, it has been proved in the chromatic theory of graphs that 4 colours are
enough for satisfying the constraint "two adjacent countries must always have
different colours"). Every time the user inserts the colour for a country, the
domains (colours) of the remaining countries are computed again applying the
constraint just entered by the user, and for the next countries the choice of
the colours is only restricted to these computed domains. If the user has to
choose a colour for a country having an only value left in its domain, that
value will be chosen automatically.

- sudoku.cpp is a nice program that automatically solves a sudoku, theoretically by
any size (in truth the complexity of the algorithm rises exponentially when the
size of the sudoku grows, and the required time for the algorithm is also
inversely proportional to the number of variables already set by default in the
game). The sudoku to be solved is read by a text file. If no file is provided,
the program reads the sudoku contained in ./sudoku.txt. More examples are
provided in files sudoku-easy.txt, sudoku-medium.txt, sudoku-hard.txt and
sudoku-2x2x2.txt. If you want to solve a sudoku different by the one provided in
sudoku.txt, just put the new sudoku in a file whose structure is similar to the
one illustrated in the sample files, and pass it as parameter for the
application (e.g. ./sudoku mysudoku.txt). The application will verify whether
the given game has a solution (unique), is indeterminate (the given variables
are not enough for reaching a unique solution) or impossible (any configuration
leads to a violation of the constraints).

For building the examples, from the root directory of the project just type
`make examples'. For removing them, type `make examples-clean'. For building
only one of them, type `make fourcolours' or `make sudoku', and specularly `make
fourcolours-clean' or `make sudoku-clean'.


LICENCE:

All the source code and examples are released under GNU GPL 3 licence.


AUTHOR:

BlackLight, <blacklight[AT]autistici[DOT]org>, http://0x00.ath.cx



Changelog
thu may 27 11:43:04 CEST 2010   BlackLight   <blacklight@autistici.org>
	* Release 0.1.1
	* Changed name method `getSize()' -> `size()'
	* Changed name method `getDomain()' -> `domain()'
	* Added `solve()' method, applying `refreshDomain()' until a fixed
	point is reached
	* doxygen regenerated

mon may 24 02:40:23 CEST 2010   BlackLight   <blacklight@autistici.org>
	* Release 0.1


sudoku.cpp
/*
 * =====================================================================================
 *
 *       Filename:  sudoku.cpp
 *
 *    Description:  A small program that using libCSP++ allows you to solve a sudoku.
 *                  The sudoku is passed as text file, an example of the format can
 *                  be found in the file sudoku.txt. If no parameters is passed to the
 *                  program, this default sudoku is loaded and solved, otherwise the
 *                  sudoku chosen by the user in the specified text file is solved.
 *
 *          Usage:  ./sudoku <text file containing the sudoku>
 *       Complile:  g++ -IPATH/TO/csp++.h -o sudoku sudoku.cpp
 *        Version:  1.0
 *        Created:  17/05/2010 09:22:25
 *       Revision:  none
 *       Compiler:  gcc
 *
 *         Author:  BlackLight (http://0x00.ath.cx), <blacklight@autistici.org>
 *        Licence:  GNU GPL v.3
 *        Company:  lulz
 *
 * =====================================================================================
 */

#include	<iostream>
#include  <fstream>
#include  <string>
#include	<cstdlib>
#include	<cmath>
#include	<csp++/csp++.h>

#define 	NOVALUE 	-1
#define 	DEFAULT_SUDOKU_FILE 	"sudoku.txt"

using namespace std;

int WIDTH  = 0;
int HEIGHT = 0;

int
toID (int i, int j)
{
	return (i * (WIDTH*WIDTH)) + j;
}

bool
rowsConstraint ( std::vector< CSPvariable<int> > variables)
{
	for ( int i=0; i < WIDTH*WIDTH; i++ ) {
		for ( int j=0; j < HEIGHT*HEIGHT; j++ ) {
			for ( int k=0; k < j; k++ ) {
				int var1 = toID(i,j);
				int var2 = toID(i,k);

				if ( variables[var1].fixed || variables[var2].fixed )  {
					if ( variables[var1].value == variables[var2].value )  {
						return false;
					}
				}
			}
		}
	}

	return true;
}

bool
colsConstraint ( std::vector< CSPvariable<int> > variables)
{
	for ( int i=0; i < HEIGHT*HEIGHT; i++ ) {
		for ( int j=0; j < WIDTH*WIDTH; j++ ) {
			for ( int k=0; k < j; k++ ) {
				int var1 = toID(j,i);
				int var2 = toID(k,i);

				if ( variables[var1].fixed || variables[var2].fixed )  {
					if ( variables[var1].value == variables[var2].value )  {
						return false;
					}
				}
			}
		}
	}

	return true;
}

bool
cellsConstraint ( std::vector< CSPvariable<int> > variables)
{
	for ( int x=0; x < WIDTH*WIDTH; x += WIDTH )  {
		for ( int y=0; y < HEIGHT*HEIGHT; y += HEIGHT )  {
			for ( int i=x; i < x+WIDTH; i++ ) {
				for ( int j=y; j < y+HEIGHT; j++ ) {
					int var1 = toID(i,j);

					for ( int k=x; k < i; k++ ) {
						for ( int l=y; l < j; l++ ) {
							int var2 = toID(k,l);

							if ( variables[var1].fixed || variables[var2].fixed )  {
								if ( variables[var1].value == variables[var2].value )  {
									return false;
								}
							}
						}
					}
				}
			}
		}
	}

	return true;
}

void
printSudoku (CSP<int> csp)
{
	for ( int i=0; i < WIDTH*WIDTH; i++ ) {
		if (!(i % WIDTH))
			cout << endl;

		for ( int j=0; j < HEIGHT*HEIGHT; j++ ) {
			int id = toID(j,i);

			if (!(j % HEIGHT))
				cout << "  ";

			if (csp.isSet(id))
				cout << csp.value(id) << " ";
			else
				cout << ". ";
		}

		cout << endl;
	}

	cout << endl;
}

CSP<int>
getSudoku (const char* sudokuFile)
{
	string line;
	vector< vector<int> > sudoku;
	ifstream in(sudokuFile);

	if (!in)
		throw "Unable to read input file";

	for ( HEIGHT = 0; getline(in, line); HEIGHT++ )  {
		vector<int> values;
		string value = "";
		bool hasNumbers = false;

		for ( int j=0; j < line.length(); j++)  {
			if (line[j] >= '0' && line[j] <= '9')  {
				value += line[j];
				hasNumbers = true;
			} else {
				if (value.length() > 0)  {
					values.push_back( atoi(value.c_str()) );
					value = "";
				}
			}
		}

		if (!hasNumbers)  {
			HEIGHT--;
			continue;
		}

		if (value.length() > 0)  {
			values.push_back( atoi(value.c_str()) );
			value = "";
		}

		if ( HEIGHT == 0 )
			WIDTH = values.size();
		else  {
			if (values.size() != WIDTH)  {
				in.close();
				throw "The columns have different number of elements";
			}
		}

		sudoku.push_back(values);
	}

	in.close();

	if (HEIGHT != WIDTH)
		throw "The grid is not a square";

	WIDTH  = (int) sqrt(WIDTH);
	HEIGHT = (int) sqrt(HEIGHT);
	vector<int> domain;

	for ( int i=1; i <= WIDTH*WIDTH; i++)
		domain.push_back(i);

	CSP<int> csp( WIDTH*WIDTH * HEIGHT*HEIGHT, NOVALUE, false);
	csp.appendConstraint(rowsConstraint);
	csp.appendConstraint(colsConstraint);
	csp.appendConstraint(cellsConstraint);

	for ( int i=0; i < csp.size(); i++ )
		csp.setDomain(i, domain);

	for ( int i=0; i < sudoku.size(); i++ ) {
		for ( int j=0; j < sudoku[i].size(); j++ ) {
			if (sudoku[i][j] != 0)  {
				csp.setValue( toID(j,i), sudoku[i][j] );
			}
		}
	}
	
	return csp;
}

int
main ( int argc, char *argv[] )
{
	string sudokuFile;
	CSP<int> csp;

	if (argc == 1)  {
		cout << "Loading default sudoku from " << DEFAULT_SUDOKU_FILE
			<< ", pass a different file as argument if you want to "
			<< "load a different sudoku\n\n";
		sudokuFile = DEFAULT_SUDOKU_FILE;
	} else {
		sudokuFile = argv[1];
	}

	try  {
		csp = getSudoku(sudokuFile.c_str());
		cout << "Sudoku to be solved:\n";
		printSudoku(csp);
	}

	catch (const char* msg)  {
		cerr << "Exception: " << msg << endl;
		return EXIT_FAILURE;
	}

	cout << "Solving..." << endl;
	csp.solve();
	cout << endl;

	printSudoku(csp);

	if (csp.isSatisfiable())  {
		cout << "This sudoku has a solution ";

		if (csp.hasUniqueSolution())
			cout << "and the solution is unique\n";
		else
			cout << "but it wasn't possible to determine it univocally (missing values?)\n";
	} else
		cout << "This sudoku does not have any solution\n";

	return EXIT_SUCCESS;
}



csp++/csp++.cpp
/*
 * =====================================================================================
 *
 *       Filename:  csp++.cpp
 *
 *    Description:  
 *
 *        Version:  0.1.1
 *        Created:  17/05/2010 09:17:13
 *       Revision:  none
 *       Compiler:  gcc
 *
 *         Author:  BlackLight (http://0x00.ath.cx), <blacklight@autistici.org>
 *        Licence:  GNU GPL v.3
 *        Company:  lulz
 *
 * =====================================================================================
 */


#include	<algorithm>

#define   __CSPPP_CPP
#include	"csp++-def.h"
#undef    __CSPPP_CPP

using std::vector;


template<class T>
void
CSP<T>::__init (int n, bool (*c)(vector< CSPvariable<T> >))
{
	variables = vector< CSPvariable<T> >(n);
	__default_domains = vector< vector<T> >(n);

	for (size_t i=0; i < variables.size(); i++)  {
		variables[i].index = i;
		variables[i].fixed = false;
	}

	constraints = vector< bool (*)(std::vector< CSPvariable<T> >) >(1);
	constraint = c;
	__has_default_value = false;
}

template<class T>
CSP<T>::CSP (int n, bool (*c)(vector< CSPvariable<T> >))
{
	__init (n, c);
}

template<class T>
CSP<T>::CSP ( int n, T default_value, bool set_value, bool (*c)(std::vector< CSPvariable<T> >) )
{
	__init (n, c);

	for ( size_t i=0; i < variables.size(); i++ )  {
		variables[i].value = default_value;
		variables[i].fixed = set_value;
	}

	__default_value = default_value;
	__has_default_value = true;
}

template<class T>
void
CSP<T>::setDomain (size_t index, vector<T> domain)
{
	if (index >= variables.size())
		throw CSPexception("Index out of range");

	variables[index].domain = domain;
	__default_domains[index] = domain;
}

template<class T>
void
CSP<T>::setDomain (size_t index, T domain[], int size)
{
	if (index >= variables.size())
		throw CSPexception("Index out of range");

	if (size < 0)
		throw CSPexception("Invalid domain size");

	variables[index].domain = vector<T>(size);

	for (int i=0; i < size; i++)  {
		variables[index].domain[i] = domain[i];
		__default_domains[index].push_back(domain[i]);
	}
}

template<class T>
void
CSP<T>::setConstraint ( bool (*c)(vector< CSPvariable<T> >))
{
	constraints = vector< bool(*)(vector< CSPvariable<T > >) >(1);
	constraint = c;
}

template<class T>
void
CSP<T>::setConstraint ( std::vector< bool(*)(std::vector< CSPvariable<T> >) > c )
{
	constraints = c;
}

template<class T>
void
CSP<T>::appendConstraint ( bool (*c)(vector< CSPvariable<T> >))
{
	constraints.push_back(c);
}

template<class T>
void
CSP<T>::dropConstraint ( size_t index )
{
	if (index >= constraints.size())
		throw CSPexception("Index out of range");

	constraints.erase( constraints.begin() + index );
}

template<class T>
void
CSP<T>::restoreDomains ( void )
{
	for (int i=0; i < variables.size(); i++)
		variables[i].domain = __default_domains[i];
}

template<class T>
void
CSP<T>::refreshDomains ( void )
{
	vector< arc<T> > arcs;
	restoreDomains();

	for (typename vector< CSPvariable<T> >::iterator x = variables.begin(); x != variables.end(); x++)  {
		for (typename vector< CSPvariable<T> >::iterator y = variables.begin(); y != variables.end(); y++)  {
			T xOrigValue = x->value;
			T yOrigValue = y->value;

			for (size_t i=0; i < x->domain.size(); i++)  {
				if ( x->fixed )  {
					if ( x->domain[i] != x->value )
						continue;
				}

				x->value = x->domain[i];

				for (size_t j=0; j < y->domain.size(); j++)  {
					if ( y->fixed )  {
						if ( y->domain[j] != y->value )
							continue;
					}

					y->value = y->domain[j];
					bool isConfigurationValid = true;

					for ( typename std::vector< bool (*)(std::vector< CSPvariable<T> >) >::iterator c = constraints.begin();
							c != constraints.end() && isConfigurationValid;
							c++ ) {
						if (!(*c)(variables))
							isConfigurationValid = false;
					}

					if (isConfigurationValid)  {
						arc<T> a;
						a.var[0] = *x; a.var[1] = *y;
						a.value[0] = x->value; a.value[1] = y->value;
						arcs.push_back(a);
					}
				}

				y->value = yOrigValue;
			}

			x->value = xOrigValue;
		}

		vector<T> domain = vector<T>();

		for (size_t i=0; i < arcs.size(); i++)  {
			if (arcs[i].var[0].index == x->index ||
					arcs[i].var[1].index == x->index)  {
				T value = (arcs[i].var[0].index == x->index) ? arcs[i].value[0] : arcs[i].value[1];
				bool found = false;

				for (size_t j=0; j < domain.size() && !found; j++)  {
					if (domain[j] == value)
						found = true;
				}

				if (!found)  {
					domain.push_back(value);
				}
			}
		}

		sort(domain.begin(), domain.end());
		x->domain = domain;
	}
}

template<class T>
std::vector<T>
CSP<T>::domain ( size_t index )
{
	if (index >= variables.size())
		throw CSPexception("Index out of range");

	return variables[index].domain;
}

template<class T>
size_t
CSP<T>::size( void )
{
	return variables.size();
}

template<class T>
void
CSP<T>::setValue ( size_t index, T value )
{
	if (index >= variables.size())
		throw CSPexception("Index out of range");

	variables[index].value = value;
	variables[index].fixed = true;
}

template<class T>
void
CSP<T>::unsetValue ( size_t index )
{
	if (index >= variables.size())
		throw CSPexception("Index out of range");

	if (__has_default_value)
		variables[index].value = __default_value;
	variables[index].fixed = false;
}

template<class T>
bool
CSP<T>::isSatisfiable ( void )
{
	for ( size_t i=0; i < variables.size(); i++ ) {
		if ( variables[i].domain.empty() )
			return false;
	}

	return true;
}

template<class T>
bool
CSP<T>::hasUniqueSolution ( void )
{
	for ( size_t i=0; i < variables.size(); i++ ) {
		if (variables[i].domain.size() != 1)
			return false;
	}

	return true;
}

template<class T>
void
CSP<T>::assignUniqueDomains ( void )
{
	for ( int i=0; i < size(); i++ )  {
		if (domain(i).size() == 1)
			setValue( i, domain(i)[0] );
	}
}

template<class T>
bool
CSP<T>::isSet ( size_t index )
{
	if (index >= variables.size())
		throw CSPexception("Index out of range");

	return variables[index].fixed;
}

template<class T>
T
CSP<T>::value ( size_t index )
{
	if (index >= variables.size())
		throw CSPexception("Index out of range");

	return variables[index].value;
}

template<class T>
void
CSP<T>::solve ( size_t max_iterations )
{
	bool   changed = false;
	size_t steps = 1;
	vector< vector<T> > oldDomains(size());

	do  {
		if (max_iterations != 0)  {
			if (steps++ > max_iterations)
				break;
		}

		for ( size_t i=0; i < size(); i++ )  {
			oldDomains[i] = variables[i].domain;
		}
		
		refreshDomains();
		assignUniqueDomains();

		if (hasUniqueSolution())
			break;

		changed = false;

		for ( size_t i=0; i < size(); i++ )  {
			if (domain(i).size() != oldDomains[i].size())
				changed = true;

			for ( size_t j=0; j < domain(i).size() && !changed; j++ ) {
				if ( domain(i)[j] != oldDomains[i][j] )
					changed = true;
			}
		}
	} while (changed);
}



fourcolours.cpp
/*
 * =====================================================================================
 *
 *       Filename:  fourcolours.cpp
 *
 *    Description:  Interactively select a colour for some European countries. At each
 *    			selection, the domains of colours for remaining countries are
 *    			restricted, under the constraint "two adjacent countries cannot have
 *    			the same colour"
 *
 *       Complile:  g++ -IPATH/TO/csp++.h -o fourcolours fourcolours.cpp
 *        Version:  1.0
 *        Created:  17/05/2010 09:22:25
 *       Revision:  none
 *       Compiler:  gcc
 *
 *         Author:  BlackLight (http://0x00.ath.cx), <blacklight@autistici.org>
 *        Licence:  GNU GPL v.3
 *        Company:  lulz
 *
 * =====================================================================================
 */

#include	<iostream>
#include  <string>
#include	<vector>
#include	<cstdlib>
#include  <cstring>
#include	<csp++/csp++.h>

using namespace std;

typedef enum  {
	I, CH, DK, D, A, F, B, E, P, L, NL,
} Country;

typedef enum  {
	red, green, blue, yellow
} Colour;

#define 	COUNTRIES 	11
#define 	COLOURS 		4

static const char* countries[COUNTRIES] = {
	"Italy", "Switz.", "Denmark", "Germany", "Austria", "France", "Belgium", "Spain", "Portugal", "Luxemb.", "Holland",
};

static const char* colours[COLOURS] = {
	"red", "green", "blue", "yellow"
};

/**
 * FUNCTION: c
 *
 * Constraint function, it expresses the logic of the constraint. In our case,
 * the adjoining countries must have different colours. Each colours condition
 * is verified only if both the values have been set ("fixed" field in CSPvariable
 * structure), in our case by the user, in order to avoid an inconstistent CSP
 * from random values found inside of the variables before the initialization
 */
bool
c ( std::vector< CSPvariable<Colour> > variables)  {
	return (
			( !(variables[I] .fixed || variables[CH].fixed) || (variables[I] .value != variables[CH].value) ) &&
			( !(variables[I] .fixed || variables[A] .fixed) || (variables[I] .value != variables[A] .value) ) &&
			( !(variables[I] .fixed || variables[F] .fixed) || (variables[I] .value != variables[F] .value) ) &&
			( !(variables[A] .fixed || variables[CH].fixed) || (variables[A] .value != variables[CH].value) ) &&
			( !(variables[A] .fixed || variables[D] .fixed) || (variables[A] .value != variables[D] .value) ) &&
			( !(variables[CH].fixed || variables[D] .fixed) || (variables[CH].value != variables[D] .value) ) &&
			( !(variables[CH].fixed || variables[F] .fixed) || (variables[CH].value != variables[F] .value) ) &&
			( !(variables[D] .fixed || variables[F] .fixed) || (variables[D] .value != variables[F] .value) ) &&
			( !(variables[E] .fixed || variables[F] .fixed) || (variables[E] .value != variables[F] .value) ) &&
			( !(variables[E] .fixed || variables[P] .fixed) || (variables[E] .value != variables[P] .value) ) &&
			( !(variables[B] .fixed || variables[F] .fixed) || (variables[B] .value != variables[F] .value) ) &&
			( !(variables[B] .fixed || variables[D] .fixed) || (variables[B] .value != variables[D] .value) ) &&
			( !(variables[B] .fixed || variables[NL].fixed) || (variables[B] .value != variables[NL].value) ) &&
			( !(variables[B] .fixed || variables[L] .fixed) || (variables[B] .value != variables[L] .value) ) &&
			( !(variables[L] .fixed || variables[F] .fixed) || (variables[L] .value != variables[F] .value) ) &&
			( !(variables[L] .fixed || variables[D] .fixed) || (variables[L] .value != variables[D] .value) ) &&
			( !(variables[NL].fixed || variables[D] .fixed) || (variables[NL].value != variables[D] .value) ) &&
			( !(variables[DK].fixed || variables[D] .fixed) || (variables[DK].value != variables[D] .value) )
		);
}

/**
 * FUNCTION: printDomain
 *
 * Given the CSP and the index of the variable, prints its allowed domain
 */
void
printDomain (CSP<Colour> csp, int variable)
{
	cout << "[ ";

	for ( size_t i=0; i < csp.domain(variable).size(); i++ ) {
		cout << colours[csp.domain(variable)[i]];

		if ( i < csp.domain(variable).size() - 1 )
			cout << ", ";
	}

	cout << " ]";
}

/**
 * FUNCTION: printDomains
 *
 * Given the CSP, prints the domains of all the variables
 */
void
printDomains (CSP<Colour> csp)
{
	for ( size_t i=0; i < csp.size(); i++)  {
		cout << "Domain for variable '" << countries[i] << "':\t";
		printDomain(csp, i);
		cout << endl;
	}

	cout << endl;

	if (csp.isSatisfiable())  {
		cout << "The CSP has a solution\n";

		if (csp.hasUniqueSolution())
			cout << "The solution is unique\n";
	} else
		cout << "The CSP does not have a solution\n";
}

/**
 * FUNCTION: getIndexByColour
 *
 * Given a colour as string, it picks up the associated index,
 * if the colour exists in the list, otherwise it throws an
 * evil exception
 */
size_t
getIndexByColour ( const char* colour )
{
	for ( size_t i=0; i < COLOURS; i++ ) {
		if (!strcmp(colours[i], colour))
			return i;
	}

	throw 666;
}

/**
 * FUNCTION: valueOK
 *
 * Given the CSP, the index of a variable and a value, it checks
 * if the given value is consistent to the domain of that variable
 */
bool
valueOK ( CSP<Colour> csp, size_t variable, Colour value )  {
	for ( size_t i=0; i < csp.domain(variable).size(); i++ ) {
		if (csp.domain(variable)[i] == value)
			return true;
	}

	return false;
}

int
main ( int argc, char *argv[] )
{
	vector<Colour> domain;

	// Create the domain containing all the allowed colours
	for ( int i=0; i < COLOURS; i++)
		domain.push_back((Colour) i);

	// The CSP will contain as many variables as the number of countries,
	// applying the logical constraint specified in "c" function
	CSP<Colour> csp(COUNTRIES, c);

	// Set the domain for the variables
	for ( size_t i=0; i < COUNTRIES; i++ )
		csp.setDomain(i, domain);

	// Repeat until we don't find a unique solution for the CSP
	while (!csp.hasUniqueSolution())  {
		for ( size_t i=0; i < COUNTRIES && !csp.hasUniqueSolution(); i++ )  {
			bool satisfiable = true;

			// If a variable has an empty domain, something went wrong
			if (csp.domain(i).size() == 0)  {
				cout << "No values left for country " << countries[i]
					<< ", the domain was probably too small (few colours) or the problem is unsatisfiable\n";
				return EXIT_FAILURE;
			}

			// If a variable has a domain containing an only element,
			// just choose it without annoying the user
			if (csp.domain(i).size() == 1)  {
				cout << "Setting colour " << colours[csp.domain(i)[0]] << " for " << countries[i] << endl;
				csp.setValue( i, csp.domain(i)[0] );
				csp.refreshDomains();
				continue;
			}

			// Repeat the loop until the user inserts a valid colour for the CSP
			do  {
				bool found;
				size_t col_index;

				// Repeat the loop until the users inserts a colour in the list
				do  {
					string colour;

					cout << "Insert a colour for country '" << countries[i] << "' between " << endl << "\t";
					printDomain(csp, i);
					cout << ": ";
					getline (cin, colour);

					try  {
						col_index = getIndexByColour(colour.c_str());
						found = true;
					}

					catch (int e)  {
						cout << "Invalid input, please try again\n";
						found = false;
					}
				} while (!found);

				// Check if the inserted colour is valid for the domain of the variable
				satisfiable = valueOK(csp, i, (Colour) col_index);

				if (!satisfiable)  {
					cout << "You made an invalid choice according to the current constraints, please try again\n";
				} else {
					csp.setValue(i, (Colour) col_index);
					csp.refreshDomains();
				}
			} while (!satisfiable);
		}
	}

	cout << endl;
	printDomains(csp);
	return EXIT_SUCCESS;
}



csp++/csp++-def.h
/*
 * =====================================================================================
 *
 *       Filename:  csp++-def.h
 *
 *    Description:  Header file containing the CSP class definition. Never include this
 *                  file directly in your sources. Include csp++.h instead.
 *
 *        Version:  0.1.1
 *        Created:  16/05/2010 23:16:42
 *       Revision:  none
 *       Compiler:  gcc
 *
 *         Author:  BlackLight (http://0x00.ath.cx), <blacklight@autistici.org>
 *        Licence:  GNU GPL v.3
 *        Company:  lulz
 *
 * =====================================================================================
 */

#ifndef __CSPPP_H
#define __CSPPP_H

#ifndef __CSPPP_CPP
#error  "csp++-def.h must not be included directly - include csp++.h instead"
#endif

#define  __CSPPP_VERSION 	"0.1.1"

#include	<vector>
#include	<exception>

/**
 * \struct CSPvariable csp++.h
 * \brief Struct used for describing a variable in the CSP
 */
template<class T>
struct CSPvariable  {
	//! Index of the variable
	int index;
	
	//! If the value of the variable is set explicitely, this value is true
	bool fixed;
	
	//! The value of the variable
	T value;
	
 	//! Domain of the variable, declared as a vector
	std::vector<T> domain;
};

/**
 * \class CSPexception csp++.h
 * \brief Class for managing exception in CSP
 */
class CSPexception : public std::exception  {
	const char* message;

public:
	/**
	 * \brief Constructor
	 * \param m Message describing the exception
	 */
	CSPexception (const char *m)  {
		message = m;
	}

	/**
	 * \brief Return the description of the exception
	 */
	virtual const char* what()  {
		return message;
	}
};

/**
 * \class CSP csp++.h
 * \brief Main class for managing a CSP
 */
template<class T>
class CSP  {
private:
	template<class TT>
	struct arc  {
		CSPvariable<TT> var[2];
		TT value[2];
	};

	std::vector< CSPvariable<T> > variables;
	std::vector< bool (*)(std::vector< CSPvariable<T> >) > constraints;
	#define 	constraint 	constraints[0]

	std::vector< std::vector<T> > __default_domains;
	T __default_value;
	bool __has_default_value;
	static bool __default_constraint ( std::vector< CSPvariable<T> > v )  { return true; }
	void __init ( int n, bool (*c)(std::vector< CSPvariable<T> >) );
	void restoreDomains ( void );

public:
	/**
	 * \brief Empty constructor - just do nothing, used for declaring an object and
	 *        initialize it later
	 */
	CSP()  {}
	
	/**
	 * \brief  Class constructor
	 * \param  n Number of variables in the CSP
	 * \param  c Boolean function representing the constraint of the CSP
	 *           If no constraint function is set in the constructor or
	 *           using the applyConstraint() method, a default function
	 *           always returning true will be used, that allows any
	 *           domain for any variable to be valid (no constraint)
	 */
	CSP ( int n, bool (*c)(std::vector< CSPvariable<T> >) = __default_constraint );

	/**
	 * \brief  Class constructor
	 * \param  n Number of variables in the CSP
	 * \param  default_value Default value for the variables in the CSP
	 *         when initialized
	 * \param  set_variables Decide whether mark the variables set with
	 *         default_value as "set" or "not set" (default: not set)
	 * \param  c Boolean function representing the constraint of the CSP
	 *           If no constraint function is set in the constructor or
	 *           using the applyConstraint() method, a default function
	 *           always returning true will be used, that allows any
	 *           domain for any variable to be valid (no constraint)
	 */
	CSP ( int n, T default_value, bool set_variables = false, bool (*c)(std::vector< CSPvariable<T> >) = __default_constraint );

	/**
	 * \brief  Set the domain for the i-th variable
	 * \param  index Variable for which we're setting the domain
	 * \param  domain Vector containing the possible values for that variable
	 */
	void setDomain ( size_t index, std::vector<T> domain );

	/**
	 * \brief  Set the domain for the i-th variable
	 * \param  index Variable for which we're setting the domain
	 * \param  domain Array containing the possible values for that variable
	 * \param  size Size of "domain" array
	 */
	void setDomain ( size_t index, T domain[], int size );

	/**
	 * \brief  Apply the constraint to the CSP as a boolean function
	 * \param  c Boolean function representing the constraint of the CSP
	 */
	void setConstraint ( bool (*c)(std::vector< CSPvariable<T> >) );

	/**
	 * \brief  Apply the constraints to the CSP as vector of boolean functions
	 * \param  c Vector containing pointers to boolean functions representing
	 *           the constraints of the CSP
	 */
	void setConstraint ( std::vector< bool(*)(std::vector< CSPvariable<T> >) > c );

	/**
	 * \brief  Drops a constraint from the CSP
	 * \param  index Index of the constraint to be dropped
	 */
	void dropConstraint ( size_t index );

	/**
	 * \brief  Append a constraint to the list of the constraint of the CSP
	 * \param  c A function pointer returning a boolean value representing the new constraint
	 */
	void appendConstraint ( bool (*c)(std::vector< CSPvariable<T> >) );

	/**
	 * \brief  Updates the domains of the variables. Any constraint or node fixed value is applied
	 * \param  max_iterations Maximum number of iterations, after which the function will quit. This
	 *         can be used to break endless loops in domains assignations.
	 */
	void refreshDomains ( void );

	/**
	 * \brief Get, if it exists, the solution of the CSP, calling refreshDomains until a fixed point
	 *        is reached
	 */
	void solve ( size_t maxIterations = 0 );

	/**
	 * \brief  Get the domain of the i-th variable
	 * \param  index Variable for which we're going to get the domain
	 * \return The domain of the i-th variable as a vector of T
	 */
	std::vector<T> domain ( size_t index );

	/**
	 * \brief  Get the number of variables in the current CSP
	 * \return Size of the CSP
	 */
	size_t size ( void );
	
	/**
	 * \brief  Set the value of a variable as a constraint
	 * \param  index Index of the parameter to be set
	 * \param  value Value to be set
	 */
	void setValue ( size_t index, T value );

	/**
	 * \brief  Marks a variable as not set, and if a default value was assigned
	 *         in the CSP constructor, this value will be set. By default, unless
	 *         specified in the constructor, all the variables are considered as
	 *         not set
	 * \param  index Index of the variable to be unset
	 */
	void unsetValue ( size_t index );

	/**
	 * \brief  Check if the current CSP, with the applied constraints, is satisfiable
	 * \return true if the CSP has at least a possible solution, false otherwise
	 */
	bool isSatisfiable ( void );

	/**
	 * \brief  Check if the CSP, with the given variables, domains and constraints,
	 *         admits a unique solution, i.e. each domain of each variable contains
	 *         an only element
	 * \return true if the CSP has an only possible solution, false otherwise
	 */
	bool hasUniqueSolution ( void );

	/**
	 * \brief  Check if any of the variables in the CSP has a domain containing an only
	 *         item. If so, it automatically assignes that item as value of the variable
	 */

	void assignUniqueDomains ( void );
	/**
	 * \brief  Check if the i-th variable in the CSP has a fixed value
	 * \param  i Index of the variable to be checked
	 * \return true if the i-th variable has a fixed value, false otherwise
	 */
	bool isSet ( size_t i );

	/**
	 * \brief  Get the value of the i-th variable of the CSP. Be careful: before
	 *         using this function you should check whether the variable you're
	 *         going to check is actually set, using the isSet() method, unless
	 *         you know what you're doing, in order to avoid reading inconsistent
	 *         data
	 * \param  i Index of the variable to get
	 * \return The value of the variable, if the variable exists
	 */
	T value ( size_t i );
};

#endif



csp++/csp++.h
/*
 * =====================================================================================
 *
 *       Filename:  csp++.h
 *
 *    Description:  
 *
 *        Version:  0.1.1
 *        Created:  23/05/2010 23:27:43
 *       Revision:  none
 *       Compiler:  gcc
 *
 *         Author:  BlackLight (http://0x00.ath.cx), <blacklight@autistici.org>
 *        Licence:  GNU GPL v.3
 *        Company:  lulz
 *
 * =====================================================================================
 */

// Yes, I know it looks stupid, but ask C++ standard developers about this.
// C++ can't expand template class code until a type is explicitely specified.
// This is nearly like not using templates at all, but join me in asking C++
// compilers developers to massively allow the export keyword and change the
// way templates are managed. Until that moment, you must always include the
// .cpp file of the library inside your source code as well, and specify later
// the types you're going to use this template class for

#include  "csp++.cpp"

